File: src/sys/dotnet/fanx/fcode/FType.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Sep 06  Andy Frank  Creation
//

using System;
using System.IO;
using Fan.Sys;

namespace Fanx.Fcode
{
  ///
  /// FType is the read fcode representation of sys::Type.
  ///
  public class FType
  {
  //////////////////////////////////////////////////////////////////////////
  // Constructor
  //////////////////////////////////////////////////////////////////////////

    public FType(FPod pod)
    {
      this.m_pod = pod;
    }

  //////////////////////////////////////////////////////////////////////////
  // Meta IO
  //////////////////////////////////////////////////////////////////////////

    public FType readMeta(FStore.Input input)
    {
      m_self  = input.u2();
      m_base  = input.u2();
      m_mixins = new int[input.u2()];
      for (int i=0; i<m_mixins.Length; i++) m_mixins[i] = input.u2();
      m_flags  = input.u4();
      return this;
    }

  //////////////////////////////////////////////////////////////////////////
  // Body IO
  //////////////////////////////////////////////////////////////////////////

    public string filename()
    {
      return "fcode/" + m_pod.typeRef(m_self).typeName + ".fcode";
    }

    public void read()
    {
      read(m_pod.m_store.read(filename()));
    }

    public void read(FStore.Input input)
    {
      if (input.fpod.m_fcodeVersion == null)
        throw new IOException("FStore.Input.version == null");

      m_fields = new FField[input.u2()];
      for (int i=0; i<m_fields.Length; i++)
        m_fields[i] = new FField().read(input);

      m_methods = new FMethod[input.u2()];
      for (int i=0; i<m_methods.Length; i++)
        m_methods[i] = new FMethod().read(input);

      m_attrs = FAttrs.read(input);

      m_hollow = false;
      input.Close();
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    public bool m_hollow = true;  // have we only read meta-data
    public FPod m_pod;            // parent pod
    public int m_self;            // self typeRef index
    public int m_flags;           // bitmask
    public int m_base;            // base typeRef index
    public int[] m_mixins;        // mixin TypeRef indexes
    public FField[] m_fields;     // fields
    public FMethod[] m_methods;   // methods
    public FAttrs m_attrs;        // type attributes
  }
}

File: src/sys/java/fanx/fcode/FType.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   26 Dec 05  Brian Frank  Creation
//
package fanx.fcode;

import java.io.*;
import java.util.*;

/**
 * FType is the fcode representation of sys::Type.
 */
public class FType
  implements FConst
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  public FType(FPod pod)
  {
    this.pod = pod;
  }

//////////////////////////////////////////////////////////////////////////
// Meta IO
//////////////////////////////////////////////////////////////////////////

  public FType readMeta(FStore.Input in) throws IOException
  {
    self   = in.u2();
    base   = in.u2();
    mixins = new int[in.u2()];
    for (int i=0; i<mixins.length; ++i) mixins[i] = in.u2();
    flags  = in.u4();
    return this;
  }

//////////////////////////////////////////////////////////////////////////
// Body IO
//////////////////////////////////////////////////////////////////////////

  public String filename()
  {
    return "fcode/" + pod.typeRef(self).typeName + ".fcode";
  }

  public void read() throws IOException
  {
    read(pod.store.read(filename()));
  }

  public void read(FStore.Input in) throws IOException
  {
    if (in == null)
      throw new IOException("Cannot read fcode: " +  pod.typeRef(self).signature);

    if (in.fpod.fcodeVersion == null)
      throw new IOException("FStore.Input.fcodeVersion == null");

    fields = new FField[in.u2()];
    for (int i=0; i<fields.length; ++i)
      fields[i] = new FField().read(in);

    methods = new FMethod[in.u2()];
    for (int i=0; i<methods.length; ++i)
      methods[i] = new FMethod().read(in);

    attrs = FAttrs.read(in);

    hollow = false;
    in.close();
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  public boolean hollow = true; // have we only read meta-data
  public FPod pod;              // parent pod
  public int self;              // self typeRef index
  public int flags;             // bitmask
  public int base;              // base typeRef index
  public int[] mixins;          // mixin TypeRef indexes
  public FField[] fields;       // fields
  public FMethod[] methods;     // methods
  public FAttrs attrs;          // type attributes

}